### PR TITLE
Corkv2 dev

### PR DESCRIPTION
**Pull request by hannydevelop**
Closes: #174 

## Description
Update Cork design. Here’s a link to the design model and specifications:

https://docs.google.com/document/d/1ptYI09U4KNP6OWsKeqNXfQJEyK2IpAR11sHu1u9Qztk/edit?usp=sharing

### CODE REVIEW

File: app/export.go [hunk 1]
```diff
@@ -46,7 +46,8 @@ func (app *SommelierApp) ExportAppStateAndValidators(
 
 // prepare for fresh start at zero height
 // NOTE zero height genesis is a temporary feature which will be deprecated
-//      in favour of export at a block height
+//
```
Comment (mvid) [inline app/export.go:49]: is this extra linebreak needed?

File: x/auction/client/cli/tx.go [hunk 2]
```diff
@@ -45,7 +45,7 @@ $ %s tx gov submit-proposal set-token-prices <path/to/proposal.json> --from=<key
 Where proposal.json contains:
 
 {
-  "title": "Best token price proposal evar",
+  "title": "Best token price proposal ever",
```
Comment (mvid) [inline x/auction/client/cli/tx.go:48]: NO FUN ALLOWED

File: x/cork/client/cli/query.go [hunk 3]
```diff
@@ -242,3 +182,5 @@ func queryScheduledBlockHeights() *cobra.Command {
 
 	return cmd
 }
+
+// TODO(bolten): add queryCorkResult and queryCorkResults
```
Comment (mvid) [inline x/cork/client/cli/query.go:186]: is this needed for initial merge?

Comment (hannydevelop) [reply x/cork/client/cli/query.go:186]: Nice catch...fixing it now

File: app/export.go (hunk 1 above)
Comment (hannydevelop) [reply app/export.go]: removed it

File: x/cork/keeper/query_server.go [hunk 4]
```diff
@@ -92,3 +62,31 @@ func (k Keeper) QueryScheduledCorksByBlockHeight(c context.Context, req *types.Q
...
 	return &response, nil
 }
+
+func (k Keeper) QueryScheduledCorksByID(c context.Context, req *types.QueryScheduledCorksByIDRequest) (*types.QueryScheduledCorksByIDResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	response := types.QueryScheduledCorksByIDResponse{}
+	response.Corks = k.GetScheduledCorksByID(ctx, req.Id)
+	return &response, nil
+}
```
Comment (cbrit) [inline x/cork/keeper/query_server.go:72]: So if I'm understanding right, this can return several sets of scheduled corks. Let's say N SPs have submitted a unique cork around the same time, so there are N sets of scheduled corks all of id 0. From a user perspective what's the value of querying id `0` and getting back more than one "scope"?

**Review by cbrit — COMMENTED** [review]
Submitting a partial review, I'll take a second pass tomorrow.

File: integration_tests/cork_test.go [hunk 5]
```diff
@@ -117,292 +114,3 @@ func (s *IntegrationTestSuite) getCurrentCount() (*sdk.Int, error) {
 
 	return &count, nil
 }
-
-func (s *IntegrationTestSuite) TestCork() {
```
[the remark this answers is not in the document]
Comment (hannydevelop) [reply integration_tests/cork_test.go:121]: Well, just like I said in sprint, we have a PR where we'll address everything concerning the tests (unit and integration). The only reason we touched that bit, is because it'll cause an error in the CI, since we don't have the initial cork structure.

File: x/cork/keeper/msg_server.go [hunk 6]
```diff
@@ -30,40 +30,6 @@ func (k Keeper) signerToValAddr(ctx sdk.Context, signer sdk.AccAddress) (sdk.Val
 	return validatorAddr, nil
 }
 
-// SubmitCork implements types.MsgServer
```
[the remark this answers is not in the document]
Comment (hannydevelop) [reply x/cork/keeper/msg_server.go:33]: Nope, we Increment the ID in the `SetScheduledCork` function, in `keeper.go`, line 52.

File: x/cork/keeper/query_server.go (hunk 4 above)
Comment (hannydevelop) [reply x/cork/keeper/query_server.go:72]: The ID is a hash of the scope and the block_height. This means that corks submitted at the same height, in the same scope, will map to the same cork ID. Therefore, it's impossible to query an ID and get corks of more than one scope.

File: x/cork/keeper/keeper.go [hunk 7]
```diff
@@ -72,136 +77,89 @@ func (k Keeper) GetCork(ctx sdk.Context, val sdk.ValAddress, contract common.Add
...
+func (k Keeper) DeleteScheduledCork(ctx sdk.Context, blockHeight uint64, id uint64, val sdk.ValAddress, contract common.Address) {
+	ctx.KVStore(k.storeKey).Delete(types.GetScheduledCorkKey(blockHeight, id, val, contract))
 }
 
-// HasCorkForContract gets the prevote for a given validator
-// CONTRACT: must provide the validator address here not the delegate address
-func (k Keeper) HasCorkForContract(ctx sdk.Context, val sdk.ValAddress, contract common.Address) bool {
-	return ctx.KVStore(k.storeKey).Has(types.GetCorkForValidatorAddressKey(val, contract))
+// IterateScheduledCorks iterates over all scheduled corks in the store
+func (k Keeper) IterateScheduledCorks(ctx sdk.Context, cb func(val sdk.ValAddress, blockHeight uint64, id uint64, cel common.Address, cork types.Cork) (stop bool)) {
```
[the remark this answers is not in the document]
Comment (hannydevelop) [reply x/cork/keeper/keeper.go:87]: good catch, legacy code, I should fix that now

File: x/cork/keeper/msg_server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cbrit) [reply x/cork/keeper/msg_server.go:33]: line 69:
```
return &types.MsgScheduleCorkResponse{}, nil
```

the ID is not being set in the response

File: x/cork/keeper/query_server.go (hunk 4 above)
Comment (cbrit) [reply x/cork/keeper/query_server.go:72]: The ID is just an `int`, and it always starts at 0 for a new scope

File: x/cork/keeper/msg_server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (hannydevelop) [reply x/cork/keeper/msg_server.go:33]: sure, that's indeed missing.

File: x/cork/keeper/query_server.go (hunk 4 above)
Comment (cbrit) [reply x/cork/keeper/query_server.go:72]: I see where I was mistaken, I think I get it now thanks!

File: x/cork/types/cork.go [hunk 8]
```diff
@@ -17,6 +18,16 @@ func (c *Cork) InvalidationScope() tmbytes.HexBytes {
 		)).Bytes()
 }
 
+func (c *Cork) IDHash(blockHeight uint64) []byte {
+	blockHeightBytes := sdk.Uint64ToBigEndian(blockHeight)
+
+	return crypto.Keccak256Hash(
+		bytes.Join(
+			[][]byte{blockHeightBytes, []byte(c.TargetContractAddress), c.EncodedContractCall},
```
Comment (cbrit) [inline x/cork/types/cork.go:26]: The point was raised in chat that we may want to convert the target contract address to a `common.Address` before converting to bytes

File: x/cork/types/proposal.go [hunk 9]
```diff
@@ -78,3 +82,37 @@ func (m *RemoveManagedCellarIDsProposal) ValidateBasic() error {
...
+func (m *ScheduledCorkProposal) ValidateBasic() error {
+	if err := govtypes.ValidateAbstract(m); err != nil {
+		return err
+	}
+
+	if len(m.ContractCallProtoJson) == 0 {
+		return fmt.Errorf("can't have an empty command")
+	}
+
+	if len(m.TargetContractAddress) == 0 {
```
Comment (cbrit) [inline x/cork/types/proposal.go:113]: Perhaps we should also convert to `common.Address` to validate the format?

**Review by cbrit — COMMENTED** [review]
A few more comments

**Comment by hannydevelop** [comment]
I'll just let you finish so I'll address everything together. Lmk when
you're done

File: x/cork/types/cork.go (hunk 8 above)
Comment (hannydevelop) [reply x/cork/types/cork.go:26]: yea, I saw that comment by Eric. Waiting on him to give the final call

File: x/cork/keeper/keeper.go [hunk 10]
```diff
@@ -291,61 +223,117 @@ func (k Keeper) IncrementInvalidationNonce(ctx sdk.Context) uint64 {
...
-		if !found {
-			corks = append(corks, cork)
-			corkPowers = append(corkPowers, validatorPower)
-		}
+func (k Keeper) IncrementScheduledCorkID(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	nextID := k.GetLatestScheduledCorkID(ctx) + 1
+	store.Set([]byte{types.LatestCorkIDKey}, sdk.Uint64ToBigEndian(nextID))
+	return nextID
+}
```
[the remark this answers is not in the document]
Comment (hannydevelop) [reply x/cork/keeper/keeper.go:268]: yep, this should be initialised at `genesis`

File: x/cork/types/cork.go (hunk 8 above)
Comment (EricBolten) [reply x/cork/types/cork.go:26]: Yes, we should make that change.

File: x/cork/client/cli/query.go [hunk 11]
```diff
@@ -91,26 +92,28 @@ func querySubmittedCorks() *cobra.Command {
...
+			req := &types.QueryScheduledCorksRequest{}
 
-			res, err := queryClient.QueryCommitPeriod(cmd.Context(), req)
+			res, err := queryClient.QueryScheduledCorks(cmd.Context(), req)
 			if err != nil {
 				return err
 			}
 
+			writer := os.Stdout
+			writer.Write([]byte(res.String()))
```
Comment (EricBolten) [inline x/cork/client/cli/query.go:116]: Why is this here?

File: x/cork/client/cli/tx.go [hunk 12]
```diff
@@ -169,3 +169,71 @@ Where proposal.json contains:
...
+$ %s tx gov submit-proposal schedule-cork <path/to/proposal.json> --from=<key_or_address>
+
+Where proposal.json contains:
+
+{
+  "title": "Dollary-doos LP Scheduled Cork Proposal",
+  "description": "I trust them, approve cork",
+  "block_height": 100000,
+  "target_contract_address": "0x123801a7D398351b8bE11C439e05C5B3259aeC9B",
+  "contract_call_proto_json": "{\"function_name\":\"cork\",\"function_args\":[]}",
```
Comment (EricBolten) [inline x/cork/client/cli/tx.go:193]: Similarly here, some more specific comments/references to the `contract_call_proto_json` are probably warranted.

File: x/cork/client/cli/tx.go [hunk 13]
```diff
@@ -169,3 +169,71 @@ Where proposal.json contains:
...
+			deposit, err := sdk.ParseCoinsNormalized(proposal.Deposit)
+			if err != nil {
+				return err
+			}
+
+			if !common.IsHexAddress(proposal.TargetContractAddress) {
+				return fmt.Errorf("%s is not a valid contract address", proposal.TargetContractAddress)
+			}
+
+			// TODO(bolten): verify properly formatted JSON? maybe put it in the ValidateBasic?
```
Comment (EricBolten) [inline x/cork/client/cli/tx.go:225]: WRT our final answers on the contract call JSON, should follow up on this as well.

File: x/cork/handler.go [hunk 14]
```diff
@@ -10,30 +10,32 @@ import (
 	"github.com/peggyjv/sommelier/v4/x/cork/types"
 )
 
-// NewHandler returns a handler for "oracle" type messages.
+// NewHandler returns a handler for "cork" type messages.
```
Comment (EricBolten) [inline x/cork/handler.go:13]: I'm not sure if the legacy handler is even required with the default GRPC handler any more.

File: x/cork/keeper/keeper_test.go [hunk 15]
```diff
@@ -37,126 +32,10 @@ func TestCellarIDs_SetGetHas(t *testing.T) {
...
-				valAddr, err := sdk.ValAddressFromBech32(valCork.Validator)
-				require.NoError(t, err)
-
-				t.Log("Verify no cork exists for the validator")
-				cork, found := k.GetCork(ctx, valAddr, sampleCellarAddr)
-				assert.Equal(t, false, found)
-				assert.Error(t, cork.ValidateBasic())
-				assert.Contains(t, cork.ValidateBasic().Error(),
-					"cork has an empty contract call body")
+func TestSetSecheduledCorkGetSecheduledCCork_Unit(t *testing.T) {
```
Comment (EricBolten) [inline x/cork/keeper/keeper_test.go:35]: With the unit tests being empty here now, we should repopulate them appropriately.

File: integration_tests/scheduled_cork_test.go [hunk 16]
```diff
@@ -129,82 +131,88 @@ func (s *IntegrationTestSuite) TestScheduledCork() {
...
+		s.Require().Len(res.Corks, 0)
+
+		s.T().Log("verify gravity contract call tx was created")
+		gbRes, err := gbClient.ContractCallTxs(context.Background(), &gbtypes.ContractCallTxsRequest{
+			Pagination: nil,
+		})
+		s.Require().NoError(err, "failed to query gravity contract call txs")
+		s.Require().Len(gbRes.Calls, 1)
+
+		s.T().Log("verify count was updated")
```
Comment (EricBolten) [inline integration_tests/scheduled_cork_test.go:215]: We should add a test for the governance version of a scheduled cork into this integration test as well.

File: x/cork/types/keys.go [hunk 17]
```diff
@@ -40,6 +39,15 @@ const (
 
 	// ScheduledCorkKeyPrefix - <prefix><block_height><val_address><address> -> <cork>
 	ScheduledCorkKeyPrefix
+
+	// LatestCorkIDKey - <key> -> uint64(latestCorkID)
+	LatestCorkIDKey
+
+	// CorkIdHashPrefix - <prefix><hash> -> uint64(id)
+	ScheduledCorkIDHashPrefix
```
Comment (EricBolten) [inline x/cork/types/keys.go:47]: I believe this is no longer used.

File: x/cork/types/proposal.go [hunk 18]
```diff
@@ -78,3 +83,37 @@ func (m *RemoveManagedCellarIDsProposal) ValidateBasic() error {
...
+func (m *ScheduledCorkProposal) ProposalType() string {
+	return ProposalTypeScheduledCork
+}
+
+func (m *ScheduledCorkProposal) ValidateBasic() error {
+	if err := govtypes.ValidateAbstract(m); err != nil {
+		return err
+	}
+
+	if len(m.ContractCallProtoJson) == 0 {
```
Comment (EricBolten) [inline x/cork/types/proposal.go:110]: Consider if we should add further validations here -- is the JSON formatted correctly? Does `HexToAddress` below return any type of error we're ignoring or does it resolve to zero if the input isn't a valid address?

**Review by EricBolten — COMMENTED** [review]
Some thoughts and comments.

File: app/export.go [hunk 19]
```diff
@@ -46,7 +46,7 @@ func (app *SommelierApp) ExportAppStateAndValidators(
 
 // prepare for fresh start at zero height
 // NOTE zero height genesis is a temporary feature which will be deprecated
-//      in favour of export at a block height
+//	in favour of export at a block height
```
[the remark this answers is not in the document]
Comment (cbrit) [reply app/export.go:49]: WSOC?

File: integration_tests/scheduled_cork_test.go [hunk 20]
```diff
@@ -129,82 +131,88 @@ func (s *IntegrationTestSuite) TestScheduledCork() {
 		currentBlockHeight := status.SyncInfo.LatestBlockHeight
 		targetBlockHeight := currentBlockHeight + 15
 
-		s.T().Log("scheduling cork calls")
+		s.T().Logf("scheduling cork calls for height %d", targetBlockHeight)
+		blockHeightBytes := sdk.Uint64ToBigEndian(uint64(targetBlockHeight))
+		corkID := hex.EncodeToString(crypto.Keccak256Hash(
```
[the remark this answers is not in the document]
Comment (cbrit) [reply integration_tests/scheduled_cork_test.go:136]: It's a method on a `Cork` and since we aren't using the gRPC cork client we never explicitly construct one, so I extracted the process. I recall trying to replace the tx client with a protoc-generated cork client but it was giving me issues.

File: proto/cork/v2/tx.proto [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+  Cork cork = 1;
+  // the block height that must be reached
+  uint64 block_height = 2;
+  // signer account address
+  string signer = 3;
+}
+
+message MsgScheduleCorkResponse {
+  // cork ID
+  string id = 1;
```
[the remark this answers is not in the document]
Comment (cbrit) [reply proto/cork/v2/tx.proto:25]: The discussion was around the fact that a response can only be offered if `BroadcastCommit` is used (because it waits for the TX to be commited to a block) by the client, and even then it has to be unwrapped. We're still returning it but I think we were going to have steward offer a way to get the hash for a particular scheduled cork?

File: x/cork/client/cli/query.go [hunk 22]
```diff
@@ -21,12 +22,12 @@ func GetQueryCmd() *cobra.Command {
 
 	corkQueryCmd.AddCommand([]*cobra.Command{
 		queryParams(),
-		querySubmittedCorks(),
-		queryCommitPeriod(),
 		queryScheduledCorks(),
 		queryCellarIDs(),
 		queryScheduledBlockHeights(),
 		queryScheduledCorksByBlockHeight(),
+		queryCorkResult(),
```
[the remark this answers is not in the document]
Comment (cbrit) [reply x/cork/client/cli/query.go:29]: Good catch. Also should probably add CLI tests like the ones in the auction module

File: x/cork/client/cli/query.go (hunk 11 above)
Comment (cbrit) [reply x/cork/client/cli/query.go:116]: I think this is an artifact of debugging that I forgot to remove.

File: proto/cork/v2/proposal.proto [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+  repeated string cellar_ids = 3;
+  string deposit = 4;
+}
+
+message ScheduledCorkProposal {
+  string title = 1;
+  string description = 2;
+  uint64 block_height = 3;
+  string target_contract_address = 4;
+  string contract_call_proto_json = 5;
```
[the remark this answers is not in the document]
Comment (cbrit) [reply proto/cork/v2/proposal.proto:42]: Doing this made me realize we probably have to use a full `ScheduleRequest` proto in this field rather than just a function call. outlined the reasoning in chat but for context:

This is what I am currently using in my dev branch, but Steward would be unable to determine what Cellar API this is in reference to if there were two with two message that had the same name/structure:

```json
{
  “trust_position”: {
    “erc20_address: “…”
  }
}
```

It's annoying because these extra fields are already in the prop, but we probably just need to make the submitter use a who `ScheduleRequest` struct to ensure Steward can decode it to the right cellar type. The code I have so far just guesses based on whether something errors, but as mentioned above it could get a false positive if we do that.

```
// whole request
{
  “cellar_id”: “…”,
  “cellar_v1”: {
    “trust_position”: {
      “erc20_address: “…”
    }
  },
  block_height: 100000
}
```

I guess we can just have Steward validate that the prop's field and json field values match.

File: app/export.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (EricBolten) [reply app/export.go:49]: Sorry, whitespace-only change.

File: proto/cork/v2/proposal.proto (hunk 23 above)
[the remark this answers is not in the document]
Comment (EricBolten) [reply proto/cork/v2/proposal.proto:59]: Given this and for the simplicity of decoding, we should probably then use a fully qualified JSON blob including the type name to do this. Bear in mind the governance proto will be a separate entity from the Submit proto, so while it will have to mirror some of its structure it can skip the extraneous information.

File: x/cork/types/proposal_test.go [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+func TestNewAddManagedCellarIDsProposal(t *testing.T) {
+	expectedMsg := &AddManagedCellarIDsProposal{
+		Title:       "Add Cellar IDs",
+		Description: "Adds Cellar IDs",
+		CellarIds:   &CellarIDSet{Ids: []string{"1", "2"}},
+	}
+
+	createdMsg := NewAddManagedCellarIDsProposal("Add Cellar IDs", "Adds Cellar IDs", &CellarIDSet{Ids: []string{"1", "2"}})
```
Comment (EricBolten) [inline x/cork/types/proposal_test.go:17]: I don't think we need to unit test constructors in this fashion, they're just meant to be simple wrappers around the struct definition.

**Review by EricBolten — COMMENTED** [review]
Some follow-up comments.

**Review by EricBolten — APPROVED** [review]
LGTM
